derive/src/options.rs

```diff
@@ -1,37 +1,63 @@
-use proc_macro::TokenStream;
+use std::str::FromStr;
+
+use crate::params_style::ParamStyle;
+use crate::rpc_attr::path_eq_str;
+
+const CLIENT_META_WORD: &str = "client";
+const SERVER_META_WORD: &str = "server";
+const PARAMS_META_KEY: &str = "params";
 
 #[derive(Debug)]
 pub struct DeriveOptions {
 	pub enable_client: bool,
 	pub enable_server: bool,
+	pub params_style: ParamStyle,
 }
 
 impl DeriveOptions {
-	pub fn new(enable_client: bool, enable_server: bool) -> Self {
+	pub fn new(enable_client: bool, enable_server: bool, params_style: ParamStyle) -> Self {
 		DeriveOptions {
 			enable_client,
 			enable_server,
+			params_style,
 		}
 	}
 
-	pub fn try_from(tokens: TokenStream) -> Result<Self, syn::Error> {
-		if tokens.is_empty() {
-			return Ok(Self::new(true, true));
-		}
-		let ident: syn::Ident = syn::parse::<syn::Ident>(tokens)?;
-		let options = {
-			let ident = ident.to_string();
-			if ident == "client" {
-				Some(Self::new(true, false))
-			} else if ident == "server" {
-				Some(Self::new(false, true))
-			} else {
-				None
+	pub fn try_from(args: syn::AttributeArgs) -> Result<Self, syn::Error> {
+		let mut options = DeriveOptions::new(false, false, ParamStyle::default());
+		for arg in args {
+			if let syn::NestedMeta::Meta(meta) = arg {
+				match meta {
+					syn::Meta::Path(p) => {
+						match p.get_ident().unwrap().to_string().as_ref() {
+							CLIENT_META_WORD => options.enable_client = true,
+							SERVER_META_WORD => options.enable_server = true,
+							_ => {}
+						};
+					}
+					syn::Meta::NameValue(nv) => {
+						if path_eq_str(&nv.path, PARAMS_META_KEY) {
+							if let syn::Lit::Str(ref lit) = nv.lit {
+								options.params_style = ParamStyle::from_str(&lit.value())
+									.map_err(|e| syn::Error::new_spanned(nv.clone(), e))?;
+							}
+						} else {
+							return Err(syn::Error::new_spanned(nv, "Unexpected RPC attribute key"));
+						}
+					}
+					_ => return Err(syn::Error::new_spanned(meta, "Unexpected use of RPC attribute macro")),
+				}
 			}
-		};
-		match options {
-			Some(options) => Ok(options),
-			None => Err(syn::Error::new(ident.span(), "Unknown attribute.")),
 		}
+		if !options.enable_client && !options.enable_server {
+			// if nothing provided default to both
+			options.enable_client = true;
+			options.enable_server = true;
+		}
+		if options.enable_server && options.params_style == ParamStyle::Named {
+			// This is not allowed at this time
+			panic!("Server code generation only supports `params = \"positional\"` at this time. This is the default setting.")
```

Comment:
I guess it also supports `params = "raw"`, no?

Reply:
Nice catch